Sources/Property.swift

```diff
@@ -687,14 +687,34 @@ public final class MutableProperty<Value>: ComposableMutablePropertyProtocol {
 	/// - returns: the result of the action.
 	@discardableResult
 	public func withValue<Result>(_ action: (Value) throws -> Result) rethrows -> Result {
-		return try atomic.withValue(action)
+		return try box.lock.sync { try action(box.value) }
 	}
 
 	deinit {
 		observer.sendCompleted()
 	}
 }
 
+/// A reference counted box which holds a recursive lock and a value storage.
+///
+/// The requirement of a `Value?` storage from composed properties prevents further
+/// implementation sharing with `MutableProperty`.
+private final class PropertyBox<Value> {
+	let lock = NSRecursiveLock()
+	var value: Value
+
+	init(_ value: Value) {
+		self.value = value
+	}
+}
```

Comment:
I'd like to see this use the same approach as `Atomic` to prevent unsafe accesses—hiding `value` and `lock` behind locking APIs.